include/wil/toolhelp32.h

```diff
@@ -0,0 +1,133 @@
+#include "resource.h"
+#ifndef __WIL_TOOLHELP32_INCLUDED
+#define __WIL_TOOLHELP32_INCLUDED
+#include <TlHelp32.h>
+#include <processthreadsapi.h>
+namespace wil
+{
+namespace details
+{
+    template <typename TEntry, typename TEnumApi, typename TCallback>
+    void do_enum_snapshot(HANDLE handle, TEntry& entry, TEnumApi&& enumApiFirst, TEnumApi&& enumApiNext, TCallback&& callback)
+    {
+        using result_t = decltype(callback(TEntry{}));
+        bool enumResult = enumApiFirst(handle, &entry);
+        if (!enumResult)
+            return;
+
+        do
+        {
+            if constexpr (wistd::is_void_v<result_t>)
+            {
+                callback(entry);
+            }
+            else if constexpr (wistd::is_same_v<result_t, bool>)
+            {
+                if (callback(entry))
+                    return;
+            }
+            else
+            {
+                static_assert(
+                    [] {
+                        return false;
+                    }(),
+                    "Callback must return void or bool");
+            }
+            enumResult = enumApiNext(handle, &entry);
+        } while (enumResult);
+    }
+} // namespace details
+
+template <typename TCallback>
+void for_each_process(TCallback&& callback)
+{
+    PROCESSENTRY32 entry{};
+    entry.dwSize = sizeof(entry);
+    details::do_enum_snapshot(
+        unique_handle{CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0)}.get(),
+        entry,
+        &Process32First,
+        &Process32Next,
+        wistd::forward<TCallback>(callback));
+}
+
+template <typename TCallback>
+void for_each_thread(TCallback&& callback)
+{
+    THREADENTRY32 entry{};
+    entry.dwSize = sizeof(entry);
+    details::do_enum_snapshot(
+        unique_handle{CreateToolhelp32Snapshot(TH32CS_SNAPTHREAD, 0)}.get(),
+        entry,
+        &Thread32First,
+        &Thread32Next,
+        wistd::forward<TCallback>(callback));
+}
+
+template <typename TCallback>
+void for_each_module(TCallback&& callback, bool include32For64Bit = false)

```

Comment:
Having function arguments after callback arguments is typically very awkward to call. To get around the issue of not being able to specify default arguments that way, you can overload the function.